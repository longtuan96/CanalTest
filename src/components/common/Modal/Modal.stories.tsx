import type { Meta, StoryObj } from "@storybook/react";

import Modal from "./Modal";

const meta = {
  component: Modal,
  decorators: [
    (Story) => (
      <div className="text-white">
        <Story />
      </div>
    ),
  ],
} satisfies Meta<typeof Modal>;

export default meta;

type Story = StoryObj<typeof meta>;

export const ModalForMovie: Story = {
  args: {
    movieId: 0,
    onClose: () => {},
    variation: "movie",
  },
};

export const ModalForTrailer: Story = {
  args: {
    movieId: 653346,
    onClose: () => {},
    variation: "trailer",
  },
};
