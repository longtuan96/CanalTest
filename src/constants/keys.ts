export const QUERY_KEYS = {
  MOVIE: "movie",
  LIST: "list",
  DETAILS: "details",
  CREDITS: "credits",
  VIDEOS: "videos",
};

export const MOVIES_LIST_TYPE = {
  POPULAR: "movie/popular",
  NOW_PLAYING: "movie/now_playing",
  TOP_RATED: "movie/top_rated",
  UPCOMING: "movie/upcoming",
};
