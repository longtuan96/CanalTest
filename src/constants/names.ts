export const BUTTON = {
  WATCH_NOW: "Watch Now",
  PLAY_TRAILER: "Play Trailer",
};
