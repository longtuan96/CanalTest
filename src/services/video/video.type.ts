import { z } from "zod";
export const VideoItemSchema = z.object({
  iso_639_1: z.string(),
  iso_3166_1: z.string(),
  name: z.string(),
  key: z.string(),
  site: z.string(),
  size: z.number(),
  type: z.string(),
  official: z.boolean(),
  published_at: z.string(),
  id: z.string(),
});
export type VideoItem = z.infer<typeof VideoItemSchema>;

export const GetVideoReponseSchema = z.object({
  id: z.number(),
  results: z.array(VideoItemSchema),
});

export type GetVideoResponse = z.infer<typeof GetVideoReponseSchema>;

export const GetVideoRequestSchema = z.object({
  movieId: z.number(),
});
export type GetVideoRequest = z.infer<typeof GetVideoRequestSchema>;

export type ModalType = "trailer" | "movie";
