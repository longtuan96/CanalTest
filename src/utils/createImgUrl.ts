import MoviePlaceHolder from "../assets/landscape-placeholder-svgrepo-com.svg";
import PeoplePlaceHolder from "../assets/person-svgrepo-com.svg";

export default function createImgUrl(
  type: "people" | "movie",
  path?: string | null,
  size = "original",
) {
  if (!path) return type === "people" ? PeoplePlaceHolder : MoviePlaceHolder;
  return `${import.meta.env.VITE_THEMOVIEDB_IMG_URL}/${size}${path}`;
}
