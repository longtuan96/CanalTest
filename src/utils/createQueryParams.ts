export default (
  queryParams: Record<string, string | number | null | undefined> = {},
) => ({
  ...queryParams,
  api_key: import.meta.env.VITE_THEMOVIEDB_KEY,
});
