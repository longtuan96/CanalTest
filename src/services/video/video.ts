import { useQuery } from "@tanstack/react-query";
import validateAndParse from "../../utils/validateAndParse";
import { api } from "../api";
import {
  GetVideoReponseSchema,
  GetVideoRequest,
  GetVideoResponse,
} from "./video.type";
import { QUERY_KEYS } from "../../constants/keys";
import createQueryParams from "../../utils/createQueryParams";

const { VIDEOS } = QUERY_KEYS;
const getVideo = async (payload: GetVideoRequest) => {
  const response = await api.get<GetVideoResponse>(
    `/movie/${payload.movieId}/videos`,
    { params: createQueryParams() },
  );
  return validateAndParse(GetVideoReponseSchema, response.data);
};

export const useGetVideo = (payload: GetVideoRequest) =>
  useQuery({
    queryKey: [VIDEOS, payload],
    queryFn: () => getVideo(payload),
  });
