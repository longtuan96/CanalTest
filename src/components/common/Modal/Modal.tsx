import { XMarkIcon } from "@heroicons/react/24/solid";
import { useEffect, MouseEvent } from "react";
import getVideoId from "../../../utils/getVideoId";
import { useGetVideo } from "../../../services/video/video";
import { ModalType } from "../../../services/video/video.type";
import { ExclamationTriangleIcon } from "@heroicons/react/24/outline";
import { Loader } from "../Loader/Loader";

type ModalProps = {
  movieId: number;
  onClose: (e?: MouseEvent) => void;
  variation: ModalType;
};

export default function Modal({
  onClose,
  movieId,
  variation = "movie",
}: ModalProps) {
  useEffect(() => {
    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === "Escape") {
        onClose?.();
      }
    };
    document.addEventListener("keydown", handleEscape);
    return () => {
      document.removeEventListener("keydown", handleEscape);
    };
  }, [{ onClose }]);

  const resultVideo = useGetVideo({ movieId });

  const renderContent = () => {
    if (variation === "trailer" && movieId) {
      return resultVideo.isPending && resultVideo.data === undefined ? (
        <div
          style={{ width: "80vw", height: "80vh" }}
          className="flex-center bg-black"
        >
          <Loader />
        </div>
      ) : (
        <iframe
          style={{ width: "80vw", height: "80vh" }}
          src={`https://www.youtube-nocookie.com/embed/${getVideoId(resultVideo.data?.results)}?autoplay=1&start=0`}
          title="YouTube video player"
          allow="accelerometer; autoplay; encrypted-media;"
          referrerPolicy="strict-origin-when-cross-origin"
        ></iframe>
      );
    } else {
      return (
        <div
          className="  flex-center"
          style={{ width: "80vw", height: "80vh" }}
        >
          <div className="flex flex-col items-center">
            <ExclamationTriangleIcon className=" size-16 text-red-700" />
            <span className="capitalize">
              this area is unavailable please return!
            </span>
          </div>
        </div>
      );
    }
  };

  return (
    <div
      className="fixed inset-0 z-30 flex w-screen flex-col items-center overflow-y-auto px-4 py-8"
      role="dialog"
      aria-label="video-modal"
      aria-modal="true"
    >
      <div
        className="fixed left-0 top-0 h-screen w-screen bg-black bg-opacity-80"
        onClick={onClose}
        role="none"
      ></div>
      <div
        className="op relative z-20 flex w-fit flex-col rounded-2xl bg-zinc-900 shadow-lg transition-all"
        role="none"
      >
        <div className="sticky -top-8 z-20 flex items-center justify-center rounded-t-2xl">
          <button
            onClick={onClose}
            className="absolute right-0 top-0 m-3.5 h-8 w-8 rounded-full bg-black bg-opacity-25 p-1 shadow-md backdrop-blur transition-all hover:bg-opacity-0"
          >
            <XMarkIcon title="exit popup button" />
          </button>
        </div>
        <div>{renderContent()}</div>
      </div>
    </div>
  );
}
