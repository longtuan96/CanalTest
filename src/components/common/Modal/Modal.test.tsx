import { describe, expect, test, vi } from "vitest";
import Modal from "./Modal";
import { TEST_MOVIE_ID } from "../../../test/constants";
import { getTree } from "../../../test/utils";
import { videoMockData } from "../../../services/video/video.mock";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

describe("Modal", () => {
  test("render correctly", () => {
    const queryClient = new QueryClient();

    vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(videoMockData),
      }),
    );
    expect(
      getTree(
        <QueryClientProvider client={queryClient}>
          <Modal
            onClose={() => {}}
            movieId={TEST_MOVIE_ID}
            variation="trailer"
          />
        </QueryClientProvider>,
      ),
    ).toMatchSnapshot();
  });
});
