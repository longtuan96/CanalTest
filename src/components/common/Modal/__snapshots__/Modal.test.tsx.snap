// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > render correctly 1`] = `
<div
  aria-label="video-modal"
  aria-modal="true"
  className="fixed inset-0 z-30 flex w-screen flex-col items-center overflow-y-auto px-4 py-8"
  role="dialog"
>
  <div
    className="fixed left-0 top-0 h-screen w-screen bg-black bg-opacity-80"
    onClick={[Function]}
    role="none"
  />
  <div
    className="op relative z-20 flex w-fit flex-col rounded-2xl bg-zinc-900 shadow-lg transition-all"
    role="none"
  >
    <div
      className="sticky -top-8 z-20 flex items-center justify-center rounded-t-2xl"
    >
      <button
        className="absolute right-0 top-0 m-3.5 h-8 w-8 rounded-full bg-black bg-opacity-25 p-1 shadow-md backdrop-blur transition-all hover:bg-opacity-0"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          data-slot="icon"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            exit popup button
          </title>
          <path
            clipRule="evenodd"
            d="M5.47 5.47a.75.75 0 0 1 1.06 0L12 10.94l5.47-5.47a.75.75 0 1 1 1.06 1.06L13.06 12l5.47 5.47a.75.75 0 1 1-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 0 1-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 0 1 0-1.06Z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </div>
    <div>
      <div
        className="flex-center bg-black"
        style={
          {
            "height": "80vh",
            "width": "80vw",
          }
        }
      >
        <div
          className="loader"
        />
      </div>
    </div>
  </div>
</div>
`;
